mdonatello/mdonatello.py

```diff
@@ -9,3 +9,138 @@
+        self.physiochem_props_checkbox.observe(self.update_display, names="value")
+        self.hbond_props_checkbox.observe(self.update_display, names="value")
+
+    def get_hbond_description(self):
+        num_h_donors = Descriptors.NumHDonors(self.ag_noh)
+        num_h_acceptors = Descriptors.NumHAcceptors(self.ag_noh)
+        return f"Show H-Bond Donors/Acceptors"
+
+    def display_molecule(self, ag, show_atom_indices, highlight_aromatic):
+        if highlight_aromatic:
+            hit_ats = [atom.GetIdx() for atom in ag.GetAtoms() if atom.GetIsAromatic()]
+            hit_bonds = [
+                bond.GetIdx() for bond in ag.GetBonds() 
+                if bond.GetBeginAtom().GetIsAromatic() and bond.GetEndAtom().GetIsAromatic()
+            ]
+            d = rdMolDraw2D.MolDraw2DSVG(300, 300)
+            d.drawOptions().addStereoAnnotation = True
+            rdMolDraw2D.PrepareAndDrawMolecule(
+                d, ag, highlightAtoms=hit_ats, highlightBonds=hit_bonds,
+                highlightAtomColors={idx: (0.5, 0, 0.5) for idx in hit_ats},
+                highlightBondColors={idx: (0.5, 0, 0.5) for idx in hit_bonds}
+            )
+            d.FinishDrawing()
+            svg = d.GetDrawingText()
+            return HTML(svg)
+        
+        if show_atom_indices:
+            for atom in ag.GetAtoms():
+                atom.SetProp("atomNote", str(atom.GetIdx()))
+        else:
+            for atom in ag.GetAtoms():
+                atom.ClearProp("atomNote")
+        
+        bio = BytesIO()
+        img = Draw.MolToImage(ag)
+        img.save(bio, format="png")
+        html = f'<img src="data:image/png;base64,{base64.b64encode(bio.getvalue()).decode()}" />'
+        return HTML(html)
+        
+    def update_display(self, _=None):
+        self.hbond_props_checkbox.description = self.get_hbond_description()
+        smiles = Chem.MolToSmiles(self.ag_noh)
+        
+        children = [
+            self.display_molecule(self.ag_noh, self.show_atom_indices_checkbox.value, self.highlight_aromatic_checkbox.value),
+            HTML(f"<h3 style='margin: 0;'>SMILES: {smiles}</h3>")
+        ]
+        
+        if self.physiochem_props_checkbox.value:
+            children.extend([
+                self.display_molecular_weight(self.ag_noh),
+                self.display_logp(self.ag_noh),
+                self.display_tpsa(self.ag_noh),
+                self.display_rotatable_bonds(self.ag_noh)
+            ])
+            
+        if self.hbond_props_checkbox.value:
+            children.extend([
+                self.display_num_h_donors(self.ag_noh),
+                self.display_num_h_acceptors(self.ag_noh)
+            ])
+        
+        self.output_molecule.children = children
+        
+    def display_molecular_weight(self, ag):
+        mw = Descriptors.MolWt(ag)
+        return HTML("<p style='margin: 0; margin-left: 100px;'>Molecular Weight: {:.2f} g/mol</p>".format(mw))
```

Comment:
can you use f-string formatting on these HTML calls?

Reply:
Yes can do, will adjust it

Reply:
Sounds tricky , but I think I can try to adjust it, if not I would ask it on Friday.